sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala

```diff
@@ -869,26 +869,50 @@ class Analyzer(override val catalogManager: CatalogManager)
     def apply(plan: LogicalPlan): LogicalPlan = plan.resolveOperatorsWithPruning(
       _.containsPattern(UNPIVOT), ruleId) {
 
-      // once children and ids are resolved, we can determine values, if non were given
-      case up: Unpivot if up.childrenResolved && up.ids.forall(_.resolved) && up.values.isEmpty =>
-        up.copy(values = up.child.output.diff(up.ids))
-
-      case up: Unpivot if !up.childrenResolved || !up.ids.forall(_.resolved) ||
-        up.values.isEmpty || !up.values.forall(_.resolved) || !up.valuesTypeCoercioned => up
+      // once children are resolved, we can determine values from ids and vice versa
+      // if only either is given
+      case up: Unpivot if up.childrenResolved &&
+        up.ids.exists(_.forall(_.resolved)) && up.values.isEmpty =>
+        val idAttrs = AttributeSet(up.ids.get)
```

Comment:
regarding https://github.com/apache/spark/pull/37407/files#r981213360 , I think a simple rule is, we require id/value columns to be `Attribute` only, if only one is specified and Spark needs to infer the other. Thus, users can still use arbitrary expressions as id/value columns if they specify both.

Reply:
Interesting, so only in this special case we restrict the expressions to be attributes, because this is the only case where we need this constraint.

Are we speaking Scala API only or Scala, Python and SQL syntax?

Reply:
For all APIs. SQL is special because it can't specify id columns.

Reply:
I am confused. You are saying it would be OK for PySpark `unpivot` to take expressions as long as both, `ids` and `values` are given, even though `pivot` and `pandas.melt` take only column names?

Reply:
Yea, because there is no ambiguity in doing so.